File: 0392_is_subsequence/go/solution_followup_test.go
package is_subsequence

import (
	"testing"
)

func TestSubsequenceChecker(t *testing.T) {
	checker := NewSubsequenceChecker("ahbgdc")

	tests := []struct {
		name     string
		s        string
		expected bool
	}{
		{"example 1: subsequence found", "abc", true},
		{"example 2: not a subsequence", "axc", false},
		{"empty s", "", true},
		{"s equals t", "ahbgdc", true},
		{"single char match", "a", true},
		{"match at end", "c", true},
		{"repeated char not enough", "aaa", false},
		{"non-contiguous match", "abdc", true},
		{"s longer than t", "ahbgdca", false},
	}

	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			result := checker.IsSubsequence(tc.s)
			if result != tc.expected {
				t.Errorf("IsSubsequence(%q) = %v, want %v", tc.s, result, tc.expected)
			}
		})
	}
}

func TestSubsequenceCheckerEmptyT(t *testing.T) {
	checker := NewSubsequenceChecker("")
	if !checker.IsSubsequence("") {
		t.Error("empty s should be subsequence of empty t")
	}
	if checker.IsSubsequence("a") {
		t.Error("non-empty s should not be subsequence of empty t")
	}
}

func TestSubsequenceCheckerRepeatedChars(t *testing.T) {
	checker := NewSubsequenceChecker("aabbc")
	tests := []struct {
		s        string
		expected bool
	}{
		{"abc", true},
		{"aab", true},
		{"aabbc", true},
		{"aabbcc", false},
	}
	for _, tc := range tests {
		if checker.IsSubsequence(tc.s) != tc.expected {
			t.Errorf("IsSubsequence(%q) want %v", tc.s, tc.expected)
		}
	}
}

func TestSubsequenceCheckerManyQueries(t *testing.T) {
	checker := NewSubsequenceChecker("abcdefghijklmnopqrstuvwxyz")
	tests := []struct {
		s        string
		expected bool
	}{
		{"ace", true},
		{"az", true},
		{"za", false},
		{"xyz", true},
	}
	for _, tc := range tests {
		if checker.IsSubsequence(tc.s) != tc.expected {
			t.Errorf("IsSubsequence(%q) want %v", tc.s, tc.expected)
		}
	}
}

File: 0392_is_subsequence/clojure/test/solution_followup_test.clj
(ns solution-followup-test
  (:require [clojure.test :refer [deftest is testing]]
            [solution-followup :refer [build-index-map is-subsequence-followup]]))

(def ahbgdc-map (build-index-map "ahbgdc"))

(deftest followup-test
  (testing "example 1: subsequence found"
    (is (= true (is-subsequence-followup ahbgdc-map "abc"))))
  (testing "example 2: not a subsequence"
    (is (= false (is-subsequence-followup ahbgdc-map "axc"))))
  (testing "empty s"
    (is (= true (is-subsequence-followup ahbgdc-map ""))))
  (testing "s equals t"
    (is (= true (is-subsequence-followup ahbgdc-map "ahbgdc"))))
  (testing "single char match"
    (is (= true (is-subsequence-followup ahbgdc-map "a"))))
  (testing "match at end"
    (is (= true (is-subsequence-followup ahbgdc-map "c"))))
  (testing "repeated char not enough"
    (is (= false (is-subsequence-followup ahbgdc-map "aaa"))))
  (testing "non-contiguous match"
    (is (= true (is-subsequence-followup ahbgdc-map "abdc"))))
  (testing "s longer than t"
    (is (= false (is-subsequence-followup ahbgdc-map "ahbgdca")))))

(deftest followup-empty-t-test
  (let [m (build-index-map "")]
    (testing "empty s on empty t"
      (is (= true (is-subsequence-followup m ""))))
    (testing "non-empty s on empty t"
      (is (= false (is-subsequence-followup m "a"))))))

(deftest followup-repeated-chars-test
  (let [m (build-index-map "aabbc")]
    (testing "abc"
      (is (= true (is-subsequence-followup m "abc"))))
    (testing "aab"
      (is (= true (is-subsequence-followup m "aab"))))
    (testing "aabbc"
      (is (= true (is-subsequence-followup m "aabbc"))))
    (testing "aabbcc"
      (is (= false (is-subsequence-followup m "aabbcc"))))))

(deftest followup-many-queries-test
  (let [m (build-index-map "abcdefghijklmnopqrstuvwxyz")]
    (testing "ace"
      (is (= true (is-subsequence-followup m "ace"))))
    (testing "az"
      (is (= true (is-subsequence-followup m "az"))))
    (testing "za"
      (is (= false (is-subsequence-followup m "za"))))
    (testing "xyz"
      (is (= true (is-subsequence-followup m "xyz"))))))
